src/ripple/app/rdb/backend/detail/impl/Node.cpp

```diff
@@ -266,76 +271,130 @@ saveValidatedLedger(
-                        }
-
-                        sql += txnId;
-                        sql += "','";
-                        sql += toBase58(account);
-                        sql += "',";
-                        sql += ledgerSeq;
-                        sql += ",";
-                        sql += txnSeq;
-                        sql += ")";
+                        accountTxBulk =
+                            "INSERT INTO AccountTransactions "
+                            "(TransID, Account, LedgerSeq, TxnSeq) VALUES ";
+                        firstAcctTx = false;
                     }
-                    sql += ";";
-                    JLOG(j.trace()) << "ActTx: " << sql;
-                    *db << sql;
+                    else
+                    {
+                        accountTxBulk += ",";
+                    }
+
+                    accountTxBulk += "('";
+                    accountTxBulk += txnId;
+                    accountTxBulk += "','";
+                    accountTxBulk += toBase58(account);
+                    accountTxBulk += "',";
+                    accountTxBulk += ledgerSeq;
+                    accountTxBulk += ",";
+                    accountTxBulk += txnSeq;
+                    accountTxBulk += ")";
                 }
-                else if (auto const& sleTxn = acceptedLedgerTx->getTxn();
-                         !isPseudoTx(*sleTxn))
+
+                if (accts.empty() && !isPseudoTx(*acceptedLedgerTx->getTxn()))
                 {
-                    // It's okay for pseudo transactions to not affect any
-                    // accounts.  But otherwise...
                     JLOG(j.warn()) << "Transaction in ledger " << seq
                                    << " affects no accounts";
-                    JLOG(j.warn()) << sleTxn->getJson(JsonOptions::none);
+                    JLOG(j.warn()) << acceptedLedgerTx->getTxn()->getJson(
+                        JsonOptions::none);
                 }
 
-                *db
-                    << (STTx::getMetaSQLInsertReplaceHeader() +
-                        acceptedLedgerTx->getTxn()->getMetaSQL(
-                            seq, acceptedLedgerTx->getEscMeta()) +
-                        ";");
+                // Build metadata insert
+                if (firstMeta)
+                {
+                    metadataBulk = STTx::getMetaSQLInsertReplaceHeader();
+                    firstMeta = false;
+                }
+                else
+                {
+                    metadataBulk += ",";
+                }
+
+                metadataBulk += acceptedLedgerTx->getTxn()->getMetaSQL(
+                    seq, acceptedLedgerTx->getEscMeta());
+            }
+
+            // Execute bulk inserts - this is where the performance gain
+            // happens. Instead of N individual INSERT statements per
+            // transaction type, we execute just 2 bulk statements for the
+            // entire ledger.
```

Comment:
Isn't the max length of a query set to 1Mib by default? What happens when the query is larger?

Reply:
reverted to original tested queries